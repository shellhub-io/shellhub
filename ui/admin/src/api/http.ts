import { AxiosInstance } from "axios";
import { setupInterceptorsTo } from "./interceptors";
import * as axiosTs from "./client";
import { Configuration } from "./client";
import { BaseAPI } from "./client/base";

const configuration = new Configuration();
configuration.basePath = `${window.location.protocol}//${window.location.host}`;
configuration.accessToken = localStorage.getItem("cloud_token") || "";

// eslint-disable-next-line import/no-mutable-exports
let adminApi = new axiosTs.AdminApi(configuration);
// eslint-disable-next-line import/no-mutable-exports
let cloudApi = new axiosTs.CloudApi(configuration);

export const createNewClient = () => {
  const newConfiguration = new Configuration();
  newConfiguration.basePath = `${window.location.protocol}//${window.location.host}`;
  newConfiguration.accessToken = localStorage.getItem("cloud_token") || "";
  adminApi = new axiosTs.AdminApi(newConfiguration);
  cloudApi = new axiosTs.CloudApi(newConfiguration);
  return { adminApi, cloudApi };
};

declare module "./client/base" {
  interface BaseAPI {
    getAxios(): AxiosInstance;
    getConfiguration(): Configuration | undefined;
    setConfiguration(configuration: Configuration): void;
  }
}

/** Returns the axios instance */
BaseAPI.prototype.getAxios = function getAxios(this: BaseAPI): AxiosInstance {
  return this.axios;
};

/** Returns the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.getConfiguration = function getConfiguration(this: BaseAPI): Configuration | undefined {
  return this.configuration;
};

/** Sets the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.setConfiguration = function setConfiguration(this: BaseAPI, configuration: Configuration): void {
  this.configuration = configuration;
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
setupInterceptorsTo(adminApi.axios);

export { adminApi, cloudApi };
