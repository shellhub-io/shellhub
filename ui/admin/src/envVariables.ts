/* eslint-disable import/prefer-default-export */
export const envVariables = {
  announcementsEnable: process.env.SHELLHUB_ANNOUNCEMENTS === "true",
  tinyMceKey: process.env.TINYMCE_API_KEY,
};
