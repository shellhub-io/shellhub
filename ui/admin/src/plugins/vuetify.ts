// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
