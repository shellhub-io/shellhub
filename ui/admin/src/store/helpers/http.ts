import Axios from "axios";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore //* use while route mapping is not in TypeScript
export default () => {
  const axios = Axios.create({
    baseURL: `${window.location.protocol}//${window.location.host}/admin`,
    headers: {
      Authorization: `Bearer ${localStorage.getItem("cloud_token")}`,
    },
  });

  return axios;
};
