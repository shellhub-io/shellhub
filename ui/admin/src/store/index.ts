import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
