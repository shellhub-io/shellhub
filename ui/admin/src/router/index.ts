import { createRouter, createWebHistory, NavigationGuardNext, RouteLocationNormalized } from "vue-router";
import Login from "@admin/views/Login.vue";
import Dashboard from "@admin/views/Dashboard.vue";
import Users from "@admin/views/Users.vue";
import SettingsLicense from "@admin/components/Settings/SettingsLicense.vue";
import SettingsAuthentication from "@admin/components/Settings/SettingsAuthentication.vue";
import Namespaces from "@admin/views/Namespaces.vue";
import Settings from "@admin/views/Settings.vue";
import { computed } from "vue";
import useLicenseStore from "@admin/store/modules/license";
import useLayoutStore from "@admin/store/modules/layout";
import useAuthStore from "@admin/store/modules/auth";
import { plugin as snackbar } from "@/plugins/snackbar"; // using direct plugin because inject() doesn't work outside components

const routes = [
  {
    path: "/login",
    name: "login",
    component: Login,
    meta: {
      layout: "SimpleLayout",
      requiresAuth: false,
    },
  },
  {
    path: "/",
    name: "dashboard",
    component: Dashboard,
  },
  {
    path: "/users",
    name: "users",
    component: Users,
  },
  {
    path: "/user/:id",
    name: "userDetails",
    component: () => import("@admin/views/UserDetails.vue"),
  },
  {
    path: "/devices",
    name: "devices",
    component: () => import("@admin/views/Device.vue"),
    redirect: {
      name: "listDevices",
    },
    children: [
      {
        path: "",
        name: "listDevices",
        component: () => import("@admin/views/Device.vue"),
      },
    ],
  },
  {
    path: "/device/:id",
    name: "deviceDetails",
    component: () => import("@admin/views/DeviceDetails.vue"),
  },
  {
    path: "/sessions",
    name: "sessions",
    component: () => import("@admin/views/Sessions.vue"),
  },
  {
    path: "/session/:id",
    name: "sessionDetails",
    component: () => import("@admin/views/SessionDetails.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    component: Settings,
    redirect: { name: "SettingProfile" },
    children: [
      {
        path: "authentication",
        name: "SettingAuthentication",
        component: SettingsAuthentication,
      },
      {
        path: "license",
        name: "SettingLicense",
        component: SettingsLicense,
      },
    ],
  },
  {
    path: "/firewall-rules",
    name: "firewall-rules",
    component: () => import("@admin/views/FirewallRules.vue"),
  },
  {
    path: "/firewall-rules/:id",
    name: "firewallRulesDetails",
    component: () => import("@admin/views/FirewallRulesDetails.vue"),
  },
  {
    path: "/namespaces",
    name: "namespaces",
    component: Namespaces,
  },
  {
    path: "/namespace/:id",
    name: "namespaceDetails",
    component: () => import("@admin/views/NamespaceDetails.vue"),
  },
  {
    path: "/announcements",
    name: "announcements",
    component: () => import("@admin/views/Announcements.vue"),
  },
  {
    path: "/announcement/:uuid",
    name: "announcementDetails",
    component: () => import("@admin/views/AnnouncementDetails.vue"),
  },
  {
    path: "/new-announcement",
    name: "new-announcement",
    component: () => import("@admin/views/NewAnnouncement.vue"),
  },
];

const router = createRouter({
  history: createWebHistory("/admin/"),
  routes,
});

router.beforeEach(
  async (to: RouteLocationNormalized, from: RouteLocationNormalized, next: NavigationGuardNext) => {
    const licenseStore = useLicenseStore();
    const layoutStore = useLayoutStore();
    const authStore = useAuthStore();

    const requiresAuth = to.meta.requiresAuth ?? true;

    const layout = to.meta.layout || "AppLayout";
    layoutStore.setLayout(layout as string);

    if (!authStore.isLoggedIn && requiresAuth) {
      return next({
        name: "login",
        query: { redirect: to.fullPath },
      });
    }

    if (authStore.isLoggedIn && !to.meta.requiresAuth) {
      const license = computed(() => licenseStore.getLicense);

      try {
        await licenseStore.get();

        if (license.value.expired && to.name !== "SettingLicense") {
          snackbar.showError("Your license has expired. Please update it and try again.");
          return next({ name: "SettingLicense" });
        }
      } catch {
        if (to.name !== "SettingLicense") {
          snackbar.showError("Failed to get your license info. Please check it and try again.");
          return next({ name: "SettingLicense" });
        }
      }
    }

    return next();
  },
);

export default router;
