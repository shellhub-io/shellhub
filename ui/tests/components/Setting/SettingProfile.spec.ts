import { createVuetify } from "vuetify";
import { flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, it, vi } from "vitest";
import MockAdapter from "axios-mock-adapter";
import { nextTick } from "vue";
import SettingProfile from "@/components/Setting/SettingProfile.vue";
import { namespacesApi, usersApi } from "@/api/http";
import { store, key } from "@/store";
import { router } from "@/router";
import { envVariables } from "@/envVariables";
import { SnackbarPlugin } from "@/plugins/snackbar";

type SettingProfileWrapper = VueWrapper<InstanceType<typeof SettingProfile>>;

describe("Settings Namespace", () => {
  let wrapper: SettingProfileWrapper;

  const vuetify = createVuetify();

  let mockNamespace: MockAdapter;

  let mockUser: MockAdapter;

  const members = [
    {
      id: "xxxxxxxx",
      username: "test",
      role: "owner",
    },
  ];

  const namespaceData = {
    name: "test",
    owner: "test",
    tenant_id: "fake-tenant-data",
    members,
    settings: {
      session_record: true,
    },
    max_devices: 3,
    devices_count: 3,
    created_at: "",
  };

  const authData = {
    status: "success",
    token: "",
    user: "test",
    name: "test",
    tenant: "fake-tenant-data",
    email: "test@test.com",
    recovery_email: "test2@test.com",
    id: "xxxxxxxx",
    auth_methods: ["local", "saml"],
    role: "owner",
    mfa: {
      enable: false,
      validate: false,
    },
  };

  const session = true;

  beforeEach(async () => {
    window.matchMedia = vi.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    }));

    vi.useFakeTimers();
    localStorage.setItem("tenant", "fake-tenant-data");
    envVariables.isCloud = true;
    envVariables.isCommunity = true; // used in the MFA switch test only

    mockNamespace = new MockAdapter(namespacesApi.getAxios());
    mockUser = new MockAdapter(usersApi.getAxios());

    mockNamespace.onGet("http://localhost:3000/api/namespaces/fake-tenant-data").reply(200, namespaceData);
    mockUser.onGet("http://localhost:3000/api/users/security").reply(200, session);
    mockUser.onGet("http://localhost:3000/api/auth/user").reply(200, authData);

    store.commit("auth/authSuccess", authData);
    store.commit("auth/changeData", authData);
    store.commit("namespaces/setNamespace", namespaceData);
    store.commit("security/setSecurity", session);

    wrapper = mount(SettingProfile, {
      global: {
        plugins: [[store, key], vuetify, router, SnackbarPlugin],
        config: {
          errorHandler: () => { /* ignore global error handler */ },
        },
      },
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Data is defined", () => {
    expect(wrapper.vm.$data).toBeDefined();
  });

  it("Renders all data-test elements", () => {
    const dataTests = [
      "account-profile-container",
      "account-profile-card",
      "profile-header",
      "user-icon",
      "profile-title",
      "profile-subtitle",
      "edit-profile-button",
      "profile-details-list",
      "name-field",
      "name-input",
      "username-field",
      "username-input",
      "email-field",
      "email-input",
      "recovery-email-field",
      "recovery-email-input",
      "mfa-card",
      "mfa-text",
      "switch-mfa",
      "delete-account",
      "delete-account-btn",
    ];

    dataTests.forEach((dataTest) => {
      const element = wrapper.find(`[data-test="${dataTest}"]`);
      expect(element.exists()).toBe(true);
    });
  });

  it("Disables the MFA switch for Community accounts", async () => {
    const switchMfa = wrapper.find('[data-test="switch-mfa"] input');
    expect(switchMfa.attributes().disabled).toBeDefined();
  });

  it("Successfully changes user data", async () => {
    mockUser.onPatch("http://localhost:3000/api/users").reply(200);

    await wrapper.findComponent('[data-test="edit-profile-button"]').trigger("click");
    await wrapper.findComponent('[data-test="name-input"]').setValue("test");
    await wrapper.findComponent('[data-test="username-input"]').setValue("test");
    await wrapper.findComponent('[data-test="email-input"]').setValue("test@test.com");
    await wrapper.findComponent('[data-test="recovery-email-input"]').setValue("test2@test.com");
    await flushPromises();

    const changeDataSpy = vi.spyOn(store, "dispatch");
    await wrapper.findComponent('[data-test="save-changes-button"]').trigger("click");

    vi.runOnlyPendingTimers();

    await nextTick();
    await flushPromises();
    expect(changeDataSpy).toHaveBeenCalledWith("users/patchData", {
      email: "test@test.com",
      name: "test",
      recovery_email: "test2@test.com",
      username: "test",
    });
  });

  it("Fails changes user data", async () => {
    const changeUserData = {
      recovery_email: "test2@test.com",
      email: "test@test.com",
      name: "test",
      username: "test",
    };

    mockUser.onPatch("http://localhost:3000/api/users").reply(401);

    const changeDataSpy = vi.spyOn(store, "dispatch");
    await wrapper.findComponent('[data-test="edit-profile-button"]').trigger("click");
    await wrapper.findComponent('[data-test="name-input"]').setValue("test");
    await wrapper.findComponent('[data-test="username-input"]').setValue("test");
    await wrapper.findComponent('[data-test="email-input"]').setValue("test@test.com");
    await wrapper.findComponent('[data-test="recovery-email-input"]').setValue("test2@test.com");
    await flushPromises();

    await wrapper.findComponent('[data-test="save-changes-button"]').trigger("click");

    vi.runOnlyPendingTimers();

    await nextTick();
    await flushPromises();
    expect(changeDataSpy).toHaveBeenCalledWith("users/patchData", changeUserData);
  });
});
