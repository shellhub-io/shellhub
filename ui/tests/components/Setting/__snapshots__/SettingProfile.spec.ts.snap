// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings Namespace > Renders the component 1`] = `
"<div data-v-b3fa301d="" class="v-container v-container--fluid v-locale--is-ltr" data-test="account-profile-container">
  <!---->
  <!---->
  <div data-v-b3fa301d="" class="v-card v-theme--light v-card--density-default v-card--variant-flat bg-transparent" data-test="account-profile-card">
    <!---->
    <div class="v-card__loader">
      <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
        <!---->
        <div class="v-progress-linear__background"></div>
        <div class="v-progress-linear__buffer" style="width: 0%;"></div>
        <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
          <div class="v-progress-linear__indeterminate">
            <div class="v-progress-linear__indeterminate long"></div>
            <div class="v-progress-linear__indeterminate short"></div>
          </div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
    <!---->
    <div data-v-b3fa301d="" class="v-card-item">
      <!---->
      <div class="v-card-item__content">
        <!---->
        <!---->
        <div data-v-b3fa301d="" class="v-list-item v-theme--light v-list-item--density-default rounded-0 v-list-item--variant-text pa-0" data-test="profile-header">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-09753bb1="" data-v-b3fa301d="" class="v-avatar v-theme--light bg-primary v-avatar--density-default v-avatar--variant-flat border" style="width: 4rem; height: 4rem;" data-test="user-icon">
              <div data-v-09753bb1="" class="v-responsive v-img v-img--booting" data-test="gravatar-image">
                <div class="v-responsive__sizer"></div>
                <!---->
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <!---->
                </transition-stub>
                <!---->
                <!---->
                <!---->
                <!---->
              </div>
              <!----><span class="v-avatar__underlay"></span>
            </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <div class="v-list-item-title">
              <h1 data-v-b3fa301d="" data-test="profile-title">Account Profile</h1>
            </div>
            <div class="v-list-item-subtitle"><span data-v-b3fa301d="" data-test="profile-subtitle">Manage your account profile</span></div>
            <!---->
          </div>
          <div class="v-list-item__append">
            <div data-v-b3fa301d="" class="mr-4"><button data-v-b3fa301d="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-text bg-secondary border" data-test="edit-profile-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
                <!----><span class="v-btn__content" data-no-activator="">Edit Profile</span>
                <!---->
                <!---->
              </button></div>
            <div class="v-list-item__spacer"></div>
          </div>
        </div>
      </div>
      <!---->
    </div>
    <div data-v-b3fa301d="" class="v-card-text pt-4">
      <div data-v-b3fa301d="" class="v-list v-theme--light v-list--border v-list--density-default v-list--one-line v-list--rounded bg-background pa-0" tabindex="0" role="listbox" data-test="profile-details-list">
        <div data-v-b3fa301d="" class="v-card-item" style="grid-template-columns: max-content 1.5fr 2fr;">
          <div class="v-card-item__prepend"><i data-v-b3fa301d="" class="mdi-badge-account mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></div>
          <div class="v-card-item__content">
            <div class="v-card-title"><span data-v-b3fa301d="" class="text-subtitle-1" data-test="name-field">Name</span></div>
            <!---->
            <!---->
          </div>
          <div class="v-card-item__append">
            <div data-v-b3fa301d="" class="v-input v-input--horizontal v-input--density-compact v-theme--light v-locale--is-ltr v-input--dirty v-input--disabled v-input--readonly v-text-field v-input--plain-underlined" data-test="name-input">
              <!---->
              <div class="v-input__control">
                <div class="v-field v-field--active v-field--disabled v-field--dirty v-field--no-label v-field--variant-plain v-theme--light v-locale--is-ltr">
                  <div class="v-field__overlay"></div>
                  <div class="v-field__loader">
                    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                      <!---->
                      <div class="v-progress-linear__background"></div>
                      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                        <div class="v-progress-linear__indeterminate">
                          <div class="v-progress-linear__indeterminate long"></div>
                          <div class="v-progress-linear__indeterminate short"></div>
                        </div>
                      </transition-stub>
                      <!---->
                    </div>
                  </div>
                  <!---->
                  <div class="v-field__field" data-no-activator="">
                    <!---->
                    <!---->
                    <!----><input readonly="" disabled="" size="1" type="text" id="input-2" aria-describedby="input-2-messages" required="" class="v-field__input" value="test">
                    <!---->
                  </div>
                  <!---->
                  <!---->
                  <div class="v-field__outline">
                    <!---->
                    <!---->
                  </div>
                </div>
              </div>
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <hr data-v-b3fa301d="" class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
        <div data-v-b3fa301d="">
          <div data-v-b3fa301d="" class="v-card-item" style="grid-template-columns: max-content 1.5fr 2fr;">
            <div class="v-card-item__prepend"><i data-v-b3fa301d="" class="mdi-account mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></div>
            <div class="v-card-item__content">
              <div class="v-card-title"><span data-v-b3fa301d="" class="text-subtitle-1" data-test="username-field">Username</span></div>
              <!---->
              <!---->
            </div>
            <div class="v-card-item__append">
              <div data-v-b3fa301d="" class="v-input v-input--horizontal v-input--density-compact v-theme--light v-locale--is-ltr v-input--disabled v-input--readonly v-text-field v-input--plain-underlined" data-test="username-input">
                <!---->
                <div class="v-input__control">
                  <div class="v-field v-field--disabled v-field--no-label v-field--variant-plain v-theme--light v-locale--is-ltr">
                    <div class="v-field__overlay"></div>
                    <div class="v-field__loader">
                      <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                        <!---->
                        <div class="v-progress-linear__background"></div>
                        <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                        <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                          <div class="v-progress-linear__indeterminate">
                            <div class="v-progress-linear__indeterminate long"></div>
                            <div class="v-progress-linear__indeterminate short"></div>
                          </div>
                        </transition-stub>
                        <!---->
                      </div>
                    </div>
                    <!---->
                    <div class="v-field__field" data-no-activator="">
                      <!---->
                      <!---->
                      <!----><input readonly="" disabled="" size="1" type="text" id="input-4" aria-describedby="input-4-messages" required="" class="v-field__input">
                      <!---->
                    </div>
                    <!---->
                    <!---->
                    <div class="v-field__outline">
                      <!---->
                      <!---->
                    </div>
                  </div>
                </div>
                <!---->
                <!---->
              </div>
            </div>
          </div>
          <hr data-v-b3fa301d="" class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
        </div>
        <div data-v-b3fa301d="" class="v-card-item" style="grid-template-columns: max-content 1.5fr 2fr;">
          <div class="v-card-item__prepend"><i data-v-b3fa301d="" class="mdi-email mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></div>
          <div class="v-card-item__content">
            <div class="v-card-title"><span data-v-b3fa301d="" class="text-subtitle-1" data-test="email-field">Email</span></div>
            <!---->
            <!---->
          </div>
          <div class="v-card-item__append">
            <div data-v-b3fa301d="" class="v-input v-input--horizontal v-input--density-compact v-theme--light v-locale--is-ltr v-input--dirty v-input--disabled v-input--readonly v-text-field v-input--plain-underlined" data-test="email-input">
              <!---->
              <div class="v-input__control">
                <div class="v-field v-field--active v-field--disabled v-field--dirty v-field--no-label v-field--variant-plain v-theme--light v-locale--is-ltr">
                  <div class="v-field__overlay"></div>
                  <div class="v-field__loader">
                    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                      <!---->
                      <div class="v-progress-linear__background"></div>
                      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                        <div class="v-progress-linear__indeterminate">
                          <div class="v-progress-linear__indeterminate long"></div>
                          <div class="v-progress-linear__indeterminate short"></div>
                        </div>
                      </transition-stub>
                      <!---->
                    </div>
                  </div>
                  <!---->
                  <div class="v-field__field" data-no-activator="">
                    <!---->
                    <!---->
                    <!----><input readonly="" disabled="" size="1" type="text" id="input-6" aria-describedby="input-6-messages" required="" class="v-field__input" value="test@test.com">
                    <!---->
                  </div>
                  <!---->
                  <!---->
                  <div class="v-field__outline">
                    <!---->
                    <!---->
                  </div>
                </div>
              </div>
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <hr data-v-b3fa301d="" class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
        <div data-v-b3fa301d="">
          <div data-v-b3fa301d="" class="v-card-item" style="grid-template-columns: max-content 1.5fr 2fr;">
            <div class="v-card-item__prepend"><i data-v-b3fa301d="" class="mdi-email-lock mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></div>
            <div class="v-card-item__content">
              <div class="v-card-title"><span data-v-b3fa301d="" class="text-subtitle-1" data-test="recovery-email-field">Recovery Email</span></div>
              <!---->
              <!---->
            </div>
            <div class="v-card-item__append">
              <div data-v-b3fa301d="" class="v-input v-input--horizontal v-input--density-compact v-theme--light v-locale--is-ltr v-input--dirty v-input--disabled v-input--readonly v-text-field v-input--plain-underlined" data-test="recovery-email-input">
                <!---->
                <div class="v-input__control">
                  <div class="v-field v-field--active v-field--disabled v-field--dirty v-field--no-label v-field--variant-plain v-theme--light v-locale--is-ltr">
                    <div class="v-field__overlay"></div>
                    <div class="v-field__loader">
                      <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                        <!---->
                        <div class="v-progress-linear__background"></div>
                        <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                        <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                          <div class="v-progress-linear__indeterminate">
                            <div class="v-progress-linear__indeterminate long"></div>
                            <div class="v-progress-linear__indeterminate short"></div>
                          </div>
                        </transition-stub>
                        <!---->
                      </div>
                    </div>
                    <!---->
                    <div class="v-field__field" data-no-activator="">
                      <!---->
                      <!---->
                      <!----><input readonly="" disabled="" size="1" type="text" id="input-8" aria-describedby="input-8-messages" required="" class="v-field__input" value="test2@test.com">
                      <!---->
                    </div>
                    <!---->
                    <!---->
                    <div class="v-field__outline">
                      <!---->
                      <!---->
                    </div>
                  </div>
                </div>
                <!---->
                <!---->
              </div>
            </div>
          </div>
          <hr data-v-b3fa301d="" class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
          <div data-v-b3fa301d="" class="v-card-item" style="grid-template-columns: max-content 1.5fr 2fr;">
            <div class="v-card-item__prepend"><i data-v-b3fa301d="" class="mdi-key mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></div>
            <div class="v-card-item__content">
              <div class="v-card-title"><span data-v-b3fa301d="" class="text-subtitle-1">Password</span></div>
              <!---->
              <!---->
            </div>
            <div class="v-card-item__append"><button data-v-b3fa301d="" type="button" class="v-btn v-theme--light text-aprimary v-btn--density-default v-btn--size-default v-btn--variant-text"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
                <!----><span class="v-btn__content" data-no-activator="">Change Password</span>
                <!---->
                <!---->
              </button>
              <!---->
              <!---->
            </div>
          </div>
          <hr data-v-b3fa301d="" class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
          <div data-v-b3fa301d="" class="d-flex align-center justify-space-between pr-4">
            <div data-v-b3fa301d="" class="v-card v-card--disabled v-card--flat v-theme--light v-card--density-default v-card--variant-elevated bg-background w-75" tabindex="-1" data-test="mfa-card">
              <!---->
              <div class="v-card__loader">
                <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                  <!---->
                  <div class="v-progress-linear__background"></div>
                  <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                  <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                    <div class="v-progress-linear__indeterminate">
                      <div class="v-progress-linear__indeterminate long"></div>
                      <div class="v-progress-linear__indeterminate short"></div>
                    </div>
                  </transition-stub>
                  <!---->
                </div>
              </div>
              <div class="v-card-item">
                <div class="v-card-item__prepend">
                  <!----><i class="mdi-fingerprint mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true" density="default"></i>
                </div>
                <div class="v-card-item__content">
                  <div class="v-card-title"><span data-v-b3fa301d="" class="text-subtitle-1">Multi-factor Authentication</span></div>
                  <!---->
                  <!---->
                </div>
                <!---->
              </div>
              <!---->
              <div data-v-b3fa301d="" class="d-flex flex-no-wrap justify-space-between">
                <div data-v-b3fa301d="">
                  <div data-v-b3fa301d="" class="v-card-text pt-0 text-justify" data-test="mfa-text"> Enable multi-factor authentication (MFA) to add an extra layer of security to your account. You'll need to enter a one-time verification code from your preferred TOTP provider to log in. </div>
                </div>
              </div>
              <!---->
              <!----><span class="v-card__underlay"></span>
            </div>
            <div data-v-b3fa301d="" aria-describedby="v-tooltip-11" class="d-flex align-center bg-background">
              <div data-v-b3fa301d="" class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-input--disabled v-switch v-switch--inset" data-test="switch-mfa">
                <!---->
                <div class="v-input__control">
                  <div class="v-selection-control v-selection-control--disabled v-selection-control--density-default">
                    <div class="v-selection-control__wrapper">
                      <div class="v-switch__track">
                        <!---->
                        <!---->
                      </div>
                      <div class="v-selection-control__input"><input disabled="" id="switch-12" aria-disabled="true" type="checkbox" aria-describedby="switch-12-messages" value="true">
                        <div class="v-switch__thumb">
                          <transition-stub name="scale-transition" appear="false" persisted="false" css="true">
                            <!---->
                          </transition-stub>
                        </div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </div>
                <!---->
                <!---->
              </div>
              <!---->
              <!---->
              <!---->
              <!---->
            </div>
            <!--teleport start-->
            <!--teleport end-->
          </div>
        </div>
        <hr data-v-b3fa301d="" class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
        <div data-v-b3fa301d="" class="v-card-item" style="grid-template-columns: max-content 1.5fr 2fr;">
          <div class="v-card-item__prepend"><i data-v-b3fa301d="" class="mdi-delete mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></div>
          <div class="v-card-item__content">
            <div class="v-card-title"><span data-v-b3fa301d="" class="text-subtitle-1" data-test="delete-account">Delete Account</span></div>
            <!---->
            <!---->
          </div>
          <div class="v-card-item__append"><button data-v-b3fa301d="" type="button" class="v-btn v-theme--light text-error v-btn--density-default v-btn--size-default v-btn--variant-text" data-test="delete-account-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator="">Delete</span>
              <!---->
              <!---->
            </button></div>
        </div>
      </div>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;
